lisa/target.py

```diff
@@ -667,28 +667,31 @@ def freeze_userspace(self):
         """
         logger = self.get_logger()
         if 'cgroups' not in self.target.modules:
-            raise RuntimeError(
-                'Failed to freeze userspace. Ensure "cgroups" module is listed '
-                'among modules in target/test configuration')
+            raise RuntimeError('Could not freeze userspace: "cgroups" devlib module is necessary')
 
         controllers = [s.name for s in self.target.cgroups.list_subsystems()]
         if 'freezer' not in controllers:
-            logger.warning('No freezer cgroup controller on target. '
-                              'Not freezing userspace')
-            yield
-            return
+            logger.warning('Could not freeze userspace: freezer cgroup controller not available on the target')
+            cm = nullcontext
 
-        exclude = self.CRITICAL_TASKS[self.target.os]
-        logger.info('Freezing all tasks except: %s', ','.join(exclude))
-        self.target.cgroups.freeze(exclude)
+        elif not self.is_rooted:
+            logger.warning('Could not freeze userspace: target is not rooted')
+            cm = nullcontext
 
-        try:
-            yield
+        else:
+            exclude = self.CRITICAL_TASKS[self.target.os]
 
-        finally:
-            logger.info('Un-freezing userspace tasks')
-            self.target.cgroups.freeze(thaw=True)
+            @contextlib.contextmanager
+            def cm():
+                logger.info('Freezing all tasks except: %s', ','.join(exclude))
+                try:
+                    yield self.target.cgroups.freeze(exclude)
+                finally:
+                    logger.info('Un-freezing userspace tasks')
+                    self.target.cgroups.freeze(thaw=True)
 
+        with cm() as x:
```

Comment:
Following the context manager flow hurt my brain a bit but I think I got it now...

Reply:
Context-manager composition 101:
```
@contextmanager
def cm1():
    yield

@contextmanager
def cm2():
    yield

@contextmanager
def cm3():
    with cm1:
        with cm2:
           yield
```

At the end, it's not much different than chaining function calls, with `return f()` replaced by `with f(): yield`